### PR TITLE
8327818: Implement Kerberos debug with sun.security.util.Debug

### PR DESCRIPTION
**Pull request by wangweij**
Change `Krb5LoginModule` debugging to use `sun.security.util.Debug`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8327830](https://bugs.openjdk.org/browse/JDK-8327830) to be approved

### Issues
 * [JDK-8327818](https://bugs.openjdk.org/browse/JDK-8327818): Implement Kerberos debug with sun.security.util.Debug (**Enhancement** - P4)
 * [JDK-8327830](https://bugs.openjdk.org/browse/JDK-8327830): Implement Kerberos debug with sun.security.util.Debug (**CSR**)

### Reviewers
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)
 * [Sibabrata Sahoo](https://openjdk.org/census#ssahoo) (@sisahoo - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18199/head:pull/18199` \
`$ git checkout pull/18199`

Update a local copy of the PR: \
`$ git checkout pull/18199` \
`$ git pull https://git.openjdk.org/jdk.git pull/18199/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18199`

View PR using the GUI difftool: \
`$ git pr show -t 18199`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18199.diff">https://git.openjdk.org/jdk/pull/18199.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18199#issuecomment-1988916313)

### CODE REVIEW

File: src/java.security.jgss/share/classes/sun/security/jgss/spnego/SpNegoToken.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 
 import org.ietf.jgss.GSSException;
 import sun.security.jgss.GSSToken;
+import sun.security.util.Debug;
```
Comment (mcpowers) [inline src/java.security.jgss/share/classes/sun/security/jgss/spnego/SpNegoToken.java:30]: Is this really needed?

File: src/jdk.security.auth/share/classes/com/sun/security/auth/module/Krb5LoginModule.java [hunk 2]
```diff
@@ -800,20 +802,18 @@ private void promptForName(boolean getPasswdFromSharedState)
...
-            if (debug) {
-                System.out.println
+            if (debug != null) {
+                debug.println
                     ("username from shared state is " + username + "\n");
             }
+            if (debug != null) {
+                debug.println
+                        ("username from shared state is " + username + "\n");
+            }
```
Comment (mcpowers) [inline src/jdk.security.auth/share/classes/com/sun/security/auth/module/Krb5LoginModule.java:812]: Looks like lines 805-808 are copied.

File: src/jdk.security.auth/share/classes/com/sun/security/auth/module/Krb5LoginModule.java [hunk 3]
```diff
@@ -1008,12 +1008,12 @@ private Credentials renewCredentials(Credentials creds)
...
             lcreds = creds.renew();
-            if (debug)
-                System.out.println("Renewed Kerberos Ticket");
+            if (debug != null)
+                debug.println("Renewed Kerberos Ticket");
         } catch (Exception e) {
             lcreds = null;
-            if (debug)
-                System.out.println("Ticket could not be renewed : "
+            if (debug != null)
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply src/jdk.security.auth/share/classes/com/sun/security/auth/module/Krb5LoginModule.java:1011]: Yes. It's only a nit.

**Review by coffeys — COMMENTED** [review]
Looks good. Few minor comments made. 

the new Debug.of method will need to implement some logic from the decorator debug PR at https://github.com/openjdk/jdk/pull/18084  -- depending on integration order, we can tackle that as needed.

File: src/java.base/share/classes/sun/security/util/Debug.java [hunk 4]
```diff
@@ -159,6 +159,33 @@ public static Debug getInstance(String option, String prefix)
...
+     * {@snippet lang=java:
+     * Map<String,String> settings = loadLoginSettings();
+     * String property = settings.get("login");
+     * Debug debug = Debug.of("login", setting);
+     * }
+     * @param option the debug option name
+     * @param property debug setting for this option
+     * @return a new Debug object if the property is true
+     */
+    public static Debug of(String option, String property) {
```
[the remark this answers is not in the document]
Comment (coffeys) [reply src/java.base/share/classes/sun/security/util/Debug.java:180]: I see. Seems reasonable then. 

I can see how folding these krb5 debug options in under the `java.security.debug` property will make debugging easier for everyone. Hopefully, something that can be done in the near future in a follow on patch.

**Review by coffeys — APPROVED** [review]

**Review by sisahoo — APPROVED** [review]

**Pull request closed without merging**